Move retrieval of user's Reader lists to Redux

bluefuton opened the pull request:
We currently use Flux to retrieve and store the Reader lists the current user is subscribed to.

This PR moves this information across to the Redux state tree, and lays the groundwork for moving other Reader list actions to Redux too.

bluefuton commented on client/reader/sidebar/index.jsx:
```diff
@@ -49,6 +52,10 @@ const ReaderSidebar = React.createClass( {
 		// If we're browsing a tag or list, open the sidebar menu
 		this.openExpandableMenuForCurrentTagOrList();
+
+		if ( isEmpty( this.props.subscribedLists ) ) {
```
Need some extra logic here to stop us from kicking off the fetch repeatedly when the user has no lists.

blowery replied:
We probably want to add a data component to handle the query. See `components/data/query-*` for some examples

bluefuton commented:
May need to make some changes to the usage of sinon and nock based on @blowery's recent changes to the test suite.

blowery commented on client/reader/sidebar/index.jsx:
```diff
@@ -197,8 +198,9 @@ const ReaderSidebar = React.createClass( {
+				<QueryReaderLists />
 				<ReaderSidebarLists
-					lists={ this.state.lists }
+					lists={ toArray( this.props.subscribedLists ) }
```
why the toArray?

blowery commented on client/state/reader/lists/selectors.js:
```diff
@@ -0,0 +1,20 @@
+ * @return {?Object}        Reader lists
+ */
+export function getSubscribedLists( state ) {
+	return state.reader.lists.items || null;
```
ah, is this why the toArray?

bluefuton replied:
Yep. I've moved it to the selector.

blowery commented on client/state/reader/lists/test/actions.js:
```diff
@@ -0,0 +1,73 @@
+						{ ID: 413, title: 'Mango & Feijoa' }
+					]
+				} );
+		} );
```
just did a quick review on how `before` and `after` work. Wouldn't hurt to add a `nock.cleanAll` in an after.

blowery replied:
actually let's make a usingNock helper?

bluefuton replied:
Where should `usingNock` live? I've re-added the `cleanAll` in the meantime.

blowery commented on client/state/reader/lists/test/reducer.js:
```diff
@@ -0,0 +1,63 @@
+
+describe( 'reducer', () => {
+	before( () => {
+		sinon.stub( console, 'warn' );
```
why this stub?

bluefuton replied:
Pinched from the posts tests! I'll ditch it.

blowery commented on client/state/reader/reducer.js:
```diff
@@ -0,0 +1,13 @@
+import lists from './lists/reducer';
+
+export default combineReducers( {
+	lists
```
i think we might want two keys. one of list subscripions, and one of list objects? subs would just be an array of list IDs (maybe owner / slug? not sure what the best key is there)

bluefuton replied:
:+1:

bluefuton commented:
Todo:
- [ ] create two keys: one for lists, one for subscriptions to those lists (same arrangement as current Flux stores)

bluefuton commented:
We now store the list IDs a user is subscribed to under a separate key, `subscribedLists`. Just need to switch the `getSubscribedLists` selector to use it.

blowery commented:
getting a schema error on reload:
![screenshot](https://cldup.com/uISXiqkAajc/VhMeDn.png)

looks like it wants the keys on the lists to be numbers and they're strings?

bluefuton commented:
> looks like it wants the keys on the lists to be numbers and they're strings?

Ah. I think it's the ID property on the object. Looks like I need `additionalProperties` = true on the subscriptions schema, too.

blowery commented:
i think the second schema is just wrong actually. the sub list should probably just be an array?

blowery commented on client/state/reader/lists/reducer.js:
```diff
@@ -0,0 +1,92 @@
+export function subscribedLists( state = {}, action ) {
+	switch ( action.type ) {
+		case READER_LISTS_RECEIVE:
+			return Object.assign( {}, state, map( action.lists, 'ID' ) );
```
i think this should either be an array (meh. harder to insert into.) or a key-mirror?

blowery replied:
Ideally it's just a `Set` :) But then you have to serialize it to and from an Array.

blowery replied:
As it stands, you're using the array index from the map to hold each ID. that means you could have the same ID in the map more than once. expanded, it's `assign( {}, { 0: 1234 }, { 0: 345, 1: 4214, 2: 4444 } )`

So it would even overwrite any existing keys...

blowery replied:
actually ... is there any reason to do the merge? each list response is definitive I think?

blowery commented:
Clean up the schema and I think this is ready to :ship:. Works well afaict once i disable schema validation. :)

bluefuton commented:
> What do you think about the sub list just being an array?

:+1:. Changed in 2b9c560.

I've also fixed the `/read/lists` endpoint to return an integer (wpcom-r132843), which should sort out the item schema validation error.

bluefuton commented:
Looks good this end, but be good to get one last review before :ship: :)

blowery commented:
looks good. :ship: :rocket: :dollar:
